### ISSUE TITLE
(Secrets Manager): Access hosted rotation lambda role

### ISSUE THREAD

**Issue by otakusid** [aws/aws-cdk#22051]
### Describe the feature

In the case of **HostedRotation** being used for the secret (f.e. **PostgreSqlSingleUser**) there is no way to access rotation lambda IAM role ARN or Name. It would be good to have a property that returns rotation lambda **role ARN** or the ability to define rotation lambda **role Name**

### Use Case

It will be helpful in case when I need to limit access to the secret to specific principals. F.e. in the case of secret that stores credentials to the RDS database root user I want to limit access to these credentials to the list of administrator IAM roles and rotation lambda. F.e.:

```
secret.add_to_resource_policy(iam.PolicyStatement.from_json({
      'Effect': 'Deny',

      'Principal': '*',

      'Action': 'secretsmanager:*',

      'Resource': '*',

      'Condition': {
        'StringNotLike': {
          'aws:PrincipalArn': [
            'ADMIN_1_ARN',
            'ADMIN_2_ARN',
            'ROTATION_LAMBDA_ROLE_ARN'
          ]
        }
      }
    }))
```

### Proposed Solution

1. Allow to define rotation function IAM role name. F.e:

```
HostedRotation      = HostedRotation.PostgreSqlSingleUser(new SingleUserHostedRotationOptions {
  FunctionName  = "database-root-user-secret-rotation",

  FunctionRoleName  = "database-root-user-secret-rotation-role",

  ...
}),
```

2. Allow to get rotation function IAM role ARN. F.e.

```
rotation = HostedRotation.PostgreSqlSingleUser()
role_arn = rotation.RoleArn
```

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.41.0

### Environment details (OS name and version, etc.)

OS: windows; lang: python

**Comment by rix0rrr** [comment]
Makes sense.

**Comment by arvchahal** [comment]
\assign

**Comment by gudipati** [comment]
Hi @otakusid,

The [underlying Cloudformation resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/TemplateReference/aws-properties-secretsmanager-rotationschedule-hostedrotationlambda.html) for `HostedRotation` does not support providing a custom IAM role name so this is currently not possible until Cloudformation supports it.

Exposing the IAM role name does not seem to be feasible either because the IAM role of the rotation Lambda function is created by CloudFormation in a nested stack and is not directly accessible.

There are two workarounds for this until Cloudformation provides better support - 
1. You can create your own custom lambda function and IAM role and provide it to `add_rotation_schedule` using the `rotationLambda` property. This way to can customize and access the lambda & IAM role as you need.
2. You can use a custom resource to fetch the IAM role ARN -
```
from aws_cdk import custom_resources as cr

get_lambda_role = cr.AwsCustomResource(self, "GetLambdaRole",
    on_update=cr.AwsSdkCall(
        service="Lambda",
        action="getFunction",
        parameters={
            "FunctionName": "secret-rotation-lambda"
        },
        output_paths=["Configuration.Role"],
        physical_resource_id=cr.PhysicalResourceId.of("resource-id")
    ),
    policy=cr.AwsCustomResourcePolicy.from_sdk_calls(
        resources=cr.AwsCustomResourcePolicy.ANY_RESOURCE
    )
)

role_arn = get_lambda_role.get_response_field("Configuration.Role")
```

Let me know if this helps and if the issue can be closed for now

**Comment by otakusid** [comment]
@gudipati, of course, workarounds can help, but I'm not sure that this issue should be closed

Because, well, both options are custom solutions, and both options require effort to maintain. What sense does the HostedRotation make in this case?

**Comment by gudipati** [comment]
@otakusid Definitely understand your concern. Unfortunately, this issue is currently blocked until Cloudformation provides support for providing IAM role name to HostedRotation or make the IAM role name accessible from the nested stack.

**Comment by otakusid** [comment]
@gudipati  Correct me if I'm wrong, but doesn't CDK control the names of the resources? If so, it looks like we don't need to wait for CloudFormation; you can just compose the function ARN on the CDK level

![Image](https://github.com/user-attachments/assets/100c2dc2-cf41-47b3-ab6e-038f37e7b5df)
![Image](https://github.com/user-attachments/assets/7e25cd74-cccc-4a4e-ba12-47edfcd6b052)

**Comment by gudipati** [comment]
> [@gudipati](https://github.com/gudipati) Correct me if I'm wrong, but doesn't CDK control the names of the resources? If so, it looks like we don't need to wait for CloudFormation; you can just compose the function ARN on the CDK level

@otakusid, Unfortunately, the IAM role ARN connected to the Lambda function is not composable, it looks something like this - `arn:aws:iam::<account-id>:role/Stack-22051-MySecret22051-SecretsManagerRDSPostgreS-1abCdeF2gHiJ`. From what I see, the underlying Cloudformation resource `AWS::SecretsManager::RotationSchedule HostedRotationLambda` doesn't provide a way to give a custom IAM role or return the IAM role name. 

Let me know if I'm missing something. If you're interested in contributing a fix, a PR would be most welcome!
